db/versions/0019-migration.sql

```diff
@@ -0,0 +1,126 @@
+    'normal'
+  );
+
+  -- use a temporary fuction to convert taken_until.  The JS code uses `new
+  -- Date(0)` as a null value, so we convert that to NULL here
+  create function pg_temp.parse_taken_until(taken_until timestamptz) returns timestamptz
+  as $$
+      begin
+        return case taken_until
+          when '1970-01-01 00:00:00+00'::timestamptz then null
+          else taken_until
+        end;
+      end;
+  $$
+  language plpgSQL
+  strict immutable;
+
+  -- left join the tasks entities with the task_group_active_sets entities
+  -- which tracks whether a task has ever been resolved or not, and reflect
+  -- that into an `ever_resolved` boolean
+  create table tasks
+  as
+    select
+      tasks_entities.task_id as task_id,
+      (value ->> 'provisionerId')::text as provisioner_id,
+      (value ->> 'workerType')::text as worker_type,
+      (value ->> 'schedulerId')::text as scheduler_id,
+      tasks_entities.task_group_id as task_group_id,
+      entity_buf_decode(value, 'dependencies')::jsonb as dependencies,
+      (value ->> 'requires')::task_requires as requires,
+      entity_buf_decode(value, 'routes')::jsonb as routes,
+      (value ->> 'priority')::task_priority as priority,
+      (value ->> 'retries')::int as retries,
+      (value ->> 'retriesLeft')::int as retries_left,
+      (value ->> 'created')::timestamptz as created,
+      (value ->> 'deadline')::timestamptz as deadline,
+      (value ->> 'expires')::timestamptz as expires,
+      entity_buf_decode(value, 'scopes')::jsonb as scopes,
+      entity_buf_decode(value, 'payload')::jsonb as payload,
+      entity_buf_decode(value, 'metadata')::jsonb as metadata,
+      entity_buf_decode(value, 'tags')::jsonb as tags,
+      entity_buf_decode(value, 'extra')::jsonb as extra,
+      entity_buf_decode(value, 'runs')::jsonb as runs,
+      case (value ->> 'takenUntil')::timestamptz
+        when '1970-01-01 00:00:00+00'::timestamptz then null
+        else (value ->> 'takenUntil')::timestamptz
+      end as taken_until,
+      (active_tasks.task_id is null)::boolean as ever_resolved,
+      etag
+    from (
+      select
+        *,
+        partition_key as task_id,
+        uuid_to_slugid(value ->> 'taskGroupId')::text as task_group_id
+      from queue_tasks_entities
+    ) as tasks_entities
+    left join (
+      select
+        partition_key as task_group_id,
+        row_key as task_id
+      from queue_task_group_active_sets_entities
+    ) as active_tasks
+    on
+      tasks_entities.task_id = active_tasks.task_id and
+      tasks_entities.task_group_id = active_tasks.task_group_id;
+
+  alter table tasks add primary key (task_id);
+  -- this index servces the purpose of queue_task_group_members_entities
+  create index tasks_task_group_id_idx on tasks (task_group_id);
+  -- this index servces the purpose of queue_task_group_active_sets_entities
+  create index tasks_task_group_id_unresolved_idx on tasks (task_group_id)
+    where not ever_resolved;
```

Comment:
Should indexes be dropped during a downgrade?

Reply:
They are automatically dropped when the table they depend on is dropped (confirmed experimentally)